### ISSUE TITLE
[BranchPlanner] Is it possible to generate a Plan in Json format in the BranchPlanner flow?

### ISSUE THREAD

**Issue by joaoarthurv** [flux-iac/tofu-controller#1228]
Good afternoon everyone, I hope you're all doing well.

I'm working on a project to set up a GitOps workflow, and after some research, I've chosen the following tools (Flux + TFController (BranchPlanner) + Jenkins).

The idea is that when a PR is opened, the BranchPlanner is executed, and it should be possible to capture the plan in Json format. This Json-formatted plan will be captured by Jenkins, where I'll perform some validations.

While studying the tool and following the [documentation](https://flux-iac.github.io/tofu-controller/), I noticed that when I open a PR, temporary PLANs are generated in ConfigMap, along with a secret related to the PLAN and another related to the TFState. So far, from what I could tell, neither of these is in Json format.

Currently, is it possible to capture the plan in Json format in the BranchPlanner flow?

-----

I've noticed that if I merge the PR and the Terraform has the `storeReadablePlan: json` attribute enabled, a plan is created in the secret in .json format. However, since I expect this Json before the Merge, this feature doesn't fit well with what I need.

If anyone has any suggestions, they are very welcome.

I appreciate the support in advance.

-----

```
apiVersion: infra.contrib.fluxcd.io/v1alpha2
kind: Terraform
metadata:
  name: pod-test
  namespace: tf-test
spec:
  serviceAccountName: tf-test
  interval: 3m
  approvePlan: auto
  branchPlanner:
    enablePathScope: true
  destroyResourcesOnDeletion: true
  refreshBeforeApply: true
  storeReadablePlan: json
  backendConfig:
    secretSuffix: pod-test
  path: ./project-test/infra/poc-tf/tf
  sourceRef:
    kind: GitRepository
    name: tf-test
    namespace: tf-test
```

```
resource "kubernetes_pod" "example" {
  metadata {
    name = "pod-test-json-validate-pr"
    namespace = "tf-test"
  }

  spec {
    container {
      image = "nginx:latest"
      name  = "nginx-container"
    }
  }
}
```

**Comment by chanwit** [comment]
Hi @joaoarthurv

Nice to see you're using the controller + a Jenkins pipeline.

We have a feature to externally validate the tfplan via a Web hook, which could be in your Jenkins system for example. But I haven't tested this Web hook feature in the Branch Planner environment yet.


@yitsushi might have some opinions

**Comment by chanwit** [comment]
https://flux-iac.github.io/tofu-controller/use-tf-controller/with-external-webhooks/

**Comment by yitsushi** [comment]
I think we wipe out the webhook part in Branch Planner, if we don't that can cause issues as I don't think we updated that part at all for Branch Planner and I don't know if we send enough information and the recipient can distinguish calls from stable and branch planner calls. If the request contains the resource name and branch name, they may be able to distinguish the two kind of requests and should be fine, but still sounds a hacky solution.

For the json output, right now the branch planner has the human readable flag enabled hard-coded and no way to change that at the moment. If Jenkins can read secrets from kubernetes, it is possible to read data from the secret, but I think that store only the machine format and the human format.

**Comment by chanwit** [comment]
Code we're referring to is here:

https://github.com/flux-iac/tofu-controller/blob/366f7bc70bda79acac96979b6d7484a32055fdfd/internal/server/polling/terraform.go#L99-L135

I didn't see any codes that reset the webhook setting. Maybe it still would work, if you wanted to try.

**Comment by chanwit** [comment]
Hi @joaoarthurv 

Yep, you would also skip TLS verification for testing purpose or binding your cert and key to the controller pod at the location suggested by the error message.

https://github.com/flux-iac/tofu-controller/blob/366f7bc70bda79acac96979b6d7484a32055fdfd/controllers/tf_controller_webhooks.go#L115-L161

**Comment by yitsushi** [comment]
> Do you think it would be too challenging to implement a temporary Plan JSON file as output in BranchPlanner?

Kind of... not challenging, but can't call it "temporary".

## tl;dr

Without changing the CRD, the Polling service, and the Informer, we can't get that behaviour without disturbing existing functionality.

## Explanation

The code on the runner has an explicit "you can have only one of them":

https://github.com/flux-iac/tofu-controller/blob/366f7bc70bda79acac96979b6d7484a32055fdfd/runner/server_save_tfplan.go#L49-L75

That means, we would need an extra field on the resource that tells the system to use json, but in that case the json content would be sent to the PR as comment as the Informer only reads the plan from the resource where we saved and sends it to github as it is (with a little bit of sugar coating).

I think the ideal solution would be to save both human readable and json plan and set a field to the "requested" format. So if a TF resource indicates it wants a "human" plan:
 - Now: save plan in human readable format
 - My suggestion: save both json and human readable formats, and save an extra field with value "human".

And that's not a temporary quick-fix. I can't think any solutions right now that wouldn't require changing CRD.
- My suggested route wouldn't change the CRD, but I think the reason why we don't save both is simply resource limits (like #536). It feels that would cause a lot more trouble, and basically invalidates that route too.
- Adding an extra field for Branch Planner to use json output would require a CRD update and a lot of extra tests to make sure it outputs human formats is more cases. I wouldn't call it "a temporary fix".
- We could change the behaviour of the Branch Planner without changing CRD, but that may upset some people. We could say "we create a comment with a format defined on the original resource", so basically without this: https://github.com/flux-iac/tofu-controller/blob/366f7bc70bda79acac96979b6d7484a32055fdfd/internal/server/polling/terraform.go#L108. But with this, if JSON is used for automation on the main branch, all PR comments would be JSON too, which is (I think) not what most people want.

**Comment by yitsushi** [comment]
Suggestion: I would like to move that issue to a discussion under "Ideas".

That would make the discussion a bit easier with more ideas in the future. This issue is a very good discussion about a feature, but it's not actionable in this form. As a result we may file multiple issues, but the way we can add this functionality can't be determined now. We may have to think of other solutions.

Most likely we can add an extra field to the CRD here: https://github.com/flux-iac/tofu-controller/blob/366f7bc70bda79acac96979b6d7484a32055fdfd/api/v1alpha2/terraform_types.go#L275-L281. But still questionable: should we create two plan secrets (that's a big change in the runner too), or put JSON as comment in those cases (no matter how ugly it is)?

// and GitHub Issues are very hard to follow if we have more than one threads, I hope one they we can have similar to code comments/discussions but for simple comments under an issue.
